psydac/ddm/cart.py

```diff
@@ -560,6 +762,929 @@ def reduce_grid(self, global_starts, global_ends):
 
         return cart
 
+    #---------------------------------------------------------------------------
+    def change_starts_ends( self, starts, ends, parent_starts,  parent_ends):
```

Comment:
What is the purpose of this function? And will it survive the next PR?

Reply:
this function will survive in the next PR, we use this to construct spaces that are defined in part of the domain and not the whole domain, like in the case of an interface space, it only acts on `p+1` elements of the space

Reply:
I have added a doc string in [7426fa5](https://github.com/pyccel/psydac/pull/213/commits/7426fa50b3353eb00a528005c2fb2be977f6a433)

Reply:
I see. And how do you make sure that it does not try to distribute along the direction orthogonal to the interface?

Reply:
I use the sub-communicators defined along each axis and I skip the directions that I want

Reply:
This function seems quite strange to me. It is only used [here](https://github.com/pyccel/psydac/blob/5c14c7284c7e6c151b6d328777bb6f90a4d681b6/psydac/linalg/stencil.py#L335) by `StencilVectorSpace.set_interface(axis, ext, cart)`. You essentially copy `self` to a new `cart` and then manually change its `starts` and `ends`.

This means that many attributes of `cart` will be in an inconsistent state, right? If the developer uses this function to  change starts and ends in multiple directions, there will be no sub-communicator left to use as they will all be wrong.

Reply:
we only need, start, end , and the subcommunicator, we assume that we use it once for the moment, along one axis

Reply:
in the next PR, we create a cart even in the sequential case, so we use there also to create the StencilVector

Reply:
but yes, if it was used 3 times in each direction in 3D for example, there will be no subcommunicator left, and we can't communicate anything with it

Reply:
Indeed this is what I find strange. If a developer uses this function based only on the information in the docstring, most likely they will use it wrong and end up with a useless cart.

Here you want to "slice" the Cart between two locations on one axis, and then use it to communicate information along the others. Any hope that we can explain this in the docstring and put a giant "WARNING"? 😅

Reply:
> we only need, start, end , and the subcommunicator, we assume that we use it once for the moment, along one axis

I could call this function once, and change `starts` and `ends` along all directions. And the result would be a totally useless `CartDecomposition` object 😉

Reply:
that is true

Reply:
it's not really useless, we still can use it like we use the StencilVectorSpace in the sequential case

Reply:
Yes it's fine in the sequential case, but there we do not need a cart.

Reply:
we can have something like this as a docstring
""" Create a slice of the cart based on the new starts and ends.
    WARNING! this function should be used carefully,
    as it might generate errors if it was not used properly in the communication process. 
"""

Reply:
> Yes it's fine in the sequential case, but there we do not need a cart.

in the next PR we will create the cart in all cases even the sequential

Reply:
[98a43fd](https://github.com/pyccel/psydac/pull/213/commits/98a43fdff8f3741310a13f9186ef7e64a64ea747)